src/Components/Component.cpp

```diff
@@ -0,0 +1,115 @@
+// MIT License
+
+// Copyright (c) 2021 YeiTingHao ShiChengFeng
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include "Component.h"
+
+template<class queueType>
+stream::Component<queueType>::Component() {
+	task = nullptr;
+	clearData();
+}
+
+template<class queueType>
+stream::Component<queueType>::~Component() {
+	stop();
```

Comment:
You may have a race condition here, the base class destructor is executed the last, and by then, the inherited class is already destroyed, but the thread is still running and calling their virtual function, possibly causing use after free.

It is recommended that you should force your users to call stop before destroying the class. The destructor only asserts that the thread is stopped.